lukingroup/pylabnet
# packaging

Modified the package to be pip installable. See `pylabnet/README.md` for the full documentation. At a high level, one can now simply run `pip install pylabnet` and have access to `import pylabnet` statements from arbitrary scripts or IPython sessions without cloning repositories, modifying path variables, or any other shenangians. Furthermore, the `pip install pylabnet` creates executables `pylabnet.exe` and `pylabnet_proxy.exe` which perform the basic launching functionality.

At a more technical level, the magic is contained in `pylabnet/setup.py`. By running the command
```bash
python setup.py sdist bdist_wheel
```
one builds a distributable that can subsequently be uploaded to pip (with proper credentials) using
```bash
twine upload dist/*
```
`setup.py` contains all relevant information about the package. This includes the addition of a license and a version numbering scheme (the version is stored in `pylabnet.__init__.py`).

Additionally, I have depreciated explicit support for dedicated conda environments in support of pip virtual environments. This has been tested to work and is documented `pylabnet/README.md`. The advantage of this is to prevent the necessity for modification of PYTHONPATH variables to manually include the pylabnet module directory. It should also prevent conflicts with other python packages (for example, the latest version of spyder).

Recommended test strategy for *users*:
1. Create pip virtual environment
```bash
python -m venv /path/to/new/virtual/testenv
```
2. Activate environment
```bash
/path/to/new/virtual/testenv/Scripts/activate
```
3. Run the command `pip install pylabnet`
4. Test `pylabnet.exe` and `pylabnet_proxy.exe` (should be able to find them from windows start/search

Recommended test strategy for *developers*:
1. Create a **new** pip virtual environment
```bash
python -m venv /path/to/another/new/virtual/testenv
```
2. Activate environment
```bash
/path/to/another/new/virtual/testenv/Scripts/activate
```
3. Go to root pylabnet directory and run 
```bash
python setup.py develop --exclude-scripts /path/to/another/new/virtual/testenv/Scripts/activate
```
4. Test basic functionality
5. If desired, make changes and try uploading a new version to pip (e.g. v0.2.6)
    - Make changes to repository
    - Change version number in `pylabnet/__init__.py`
    - Add updates to `CHANGELOG.md`
    - commit/push
    - Create a git tag for v0.2.6 for latest commit
    - From root directory run (with environment activated) `python setup.py sdist bdist_wheel` (requires `pip install wheel`)
    - Run `twine upload dist/*` (make sure you have pip credentials)
